docs/design/datacontracts/Loader.md

```diff
@@ -0,0 +1,118 @@
+# Contract Loader
+
+This contract is for getting information about loaded modules and assemblies
+
+## APIs of contract
+
+``` csharp
+readonly struct ModuleHandle
+{
+    // Opaque handle - no public members
+
+    internal TargetPointer Address;
+}
+
+[Flags]
+enum ModuleFlags
+{
+    EditAndContinue = 0x00000008, // Edit and Continue is enabled for this module
+}
+
+enum ModuleLookupTable
+{
+    FieldDefToDesc,
+    ManifestModuleReferences,
+    MemberRefToDesc,
+    MethodDefToDesc,
+    TypeDefToMethodTable,
+    TypeRefToMethodTable,
+}
+```
+
+``` csharp
+ModuleHandle GetModuleHandle(TargetPointer);
+TargetPointer GetAssembly(ModuleHandle handle);
+ModuleFlags GetFlags(ModuleHandle handle);
+TargetPointer GetLoaderAllocator(ModuleHandle handle);
+TargetPointer GetThunkHeap(ModuleHandle handle);
+bool IsReflectionEmit(ModuleHandle handle);
+TargetPointer GetILBase(ModuleHandle handle);
+TargetPointer GetMetadataAddress(ModuleHandle handle, out ulong size);
+IDictionary<ModuleLookupTable, TargetPointer> GetLookupTables(ModuleHandle handle);
+```
+
+## Version 1
+
+Data descriptors used:
+- `Module`
+- `PEAssembly`
+
+``` csharp
+ModuleHandle GetModuleHandle(TargetPointer modulePointer)
+{
+    return new ModuleHandle(modulePointer);
+}
+
+TargetPointer GetAssembly(ModuleHandle handle)
+{
+    return target.ReadPointer(handle.Address + /* Module::Assrembly offset */);
+}
+
+ModuleFlags GetFlags(ModuleHandle handle)
+{
+    return target.Read<uint>(handle.Address + /* Module::Flags offset */);
+}
+
+TargetPointer GetLoaderAllocator(ModuleHandle handle)
+{
+    return target.ReadPointer(handle.Address + /* Module::LoaderAllocator offset */);
+}
+
+TargetPointer GetThunkHeap(ModuleHandle handle)
+{
+    return target.ReadPointer(handle.Address + /* Module::ThunkHeap offset */);
+}
+
+bool IsReflectionEmit(ModuleHandle handle)
+{
+    TargetPointer peAssembly = target.ReadPointer(handle.Address + /* Module::PEAssembly offset */);
+    TargetPointer peImage = target.ReadPointer(peAssembly + /* PEAssembly::PEImage offset */);
```

Comment:
Can we cache the reflection.emit bit on module so that we can avoid PEAssembly in the contract as well?